### PR TITLE
feat: add search bar to conversation list - WPB-7301

### PR DESCRIPTION
**Pull request by caldrian**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-7301" title="WPB-7301" target="_blank">StoryWPB-7301</a>  [iOS] Search: "Conversations" - Add search bar to top of conversation list
  </td></table>
  <br />
 

### Issue

This PR adds a search bar to the top of the conversation list.
Conversations are filtered by participant's and group conversation's names.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

### CODE REVIEW

File: setup.sh [hunk 1]
```diff
@@ -92,10 +92,6 @@ echo "ℹ️  Overriding configuration if specified..."
 scripts/override-configuration_if_needed.sh "$@"
 echo "" 
 
-echo "ℹ️  Doing additional postprocessing..."
-scripts/postprocess.sh
-echo ""
```
Comment (markusfassbender) [inline setup.sh]: Why do these changes from another PR occur here?

**Review by markusfassbender — CHANGES_REQUESTED** [review]
Something with git merge seems wrong

File: wire-ios/Wire-iOS Tests/ReferenceImages/ConversationListViewControllerTests/testForShowingConversationsFilteredByFavourites.1.png
Comment (markusfassbender) [inline wire-ios/Wire-iOS Tests/ReferenceImages/ConversationListViewControllerTests/testForShowingConversationsFilteredByFavourites.1.png:1]: This "Filtered by Favorites Remove" is now even harder to understand with the search bar in between.
We should switch to the normal Apple behaviour to tap on the highlighted filter again to remove it.

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+NavigationBar/ConversationFilterType.swift [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+enum ConversationFilterType {
+    case favorites, groups, oneToOneConversations
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+NavigationBar/ConversationFilterType.swift]: Should it be multiple lines?
```suggestion
    case favorites
    case groups
    case oneToOneConversations
```

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    func updateSearchResults(for searchController: UISearchController) {
+
+        NSObject.cancelPreviousPerformRequests(
+            withTarget: self,
+            selector: #selector(applySearchText),
+            object: nil
+        )
+
+        perform(#selector(applySearchText), with: nil, afterDelay: searchDebounceInterval)
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift:33]: Should we go to prefer Combine here over Objective-C sectors?

File: wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+public struct SearchConversationsUseCase<ConversationContainer>: SearchConversationsUseCaseProtocol
+where ConversationContainer: SearchableConversationContainer {
+
+    var conversationContainers: [ConversationContainer]
```
Comment (markusfassbender) [inline wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift]: Should the use case have a private scope for properties and just provide the `invoke` method?

File: wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCaseProtocol.swift [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+/// Determines if a user has a valid E2EI certificate on all clients.
+public protocol SearchConversationsUseCaseProtocol {
+    associatedtype ConversationContainer: SearchableConversationContainer
+
+    /// Returns `true` if all clients of the provided conversation's user have valid E2EI certificates.
+    func invoke(searchText: String) -> [ConversationContainer]
```
Comment (markusfassbender) [inline wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCaseProtocol.swift]: The comment "Returns `true`" seems outdated

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift (hunk 3 above)
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift:33]: Does this debounce or just perform after a delay? Or is it the `cancelPreviousPerformRequests` in combination with the delayed perform what achieves the debouncing?

File: wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+public struct SearchConversationsUseCase<ConversationContainer>: SearchConversationsUseCaseProtocol
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift:21]: Are there unit tests for this use case?

Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift:21]: Now there are, thanks.

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift (hunk 3 above)
Comment (caldrian) [reply wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift:33]: The combination of `cancelPreviousPerformRequests` and another `perform` results in debouncing, exactly.
We have this pattern several times in the codebase so I adopted it.

Comment (caldrian) [reply wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift:33]: Since we are doing it like this before and it comes for free with `NSObject` I prefer to keep it that way.

File: wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCaseProtocol.swift (hunk 5 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCaseProtocol.swift]: Thanks, doc comment fixed.

File: wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift (hunk 4 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift]: Ok, I made the struct immutable.

File: wire-ios/Wire-iOS Tests/ReferenceImages/ConversationListViewControllerTests/testForShowingConversationsFilteredByFavourites.1.png
Comment (caldrian) [reply wire-ios/Wire-iOS Tests/ReferenceImages/ConversationListViewControllerTests/testForShowingConversationsFilteredByFavourites.1.png:1]: Thank you for the comment.
I just realised that the positions are swapped, I'll see if it can be fixed.
Screenshot 2024-07-03 at 12 27 18 PM
Apart from that, I have to stick to the designs.

Comment (caldrian) [reply wire-ios/Wire-iOS Tests/ReferenceImages/ConversationListViewControllerTests/testForShowingConversationsFilteredByFavourites.1.png:1]: The designer now decided that it should stay like it was.

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+NavigationBar/ConversationFilterType.swift (hunk 2 above)
Comment (caldrian) [reply wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+NavigationBar/ConversationFilterType.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift (hunk 3 above)
Comment (johnxnguyen) [reply wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift:33]: if we can express it more succinctly with a Combine publisher then I would do that, it seems like a very valid use case for a publisher. It's just a suggestion though, so feel free to ignore.

**Review by johnxnguyen — APPROVED** [review]
Looks good, thank you

**Review by markusfassbender — APPROVED** [review]

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift (hunk 3 above)
Comment (caldrian) [reply wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+UISearchResultsUpdating.swift:33]: Let's do it another time.
Really, the very same code is used for searching in the start-conversation UI.
